### ISSUE TITLE
Support for an in-process server

### ISSUE THREAD

**Issue by janoc** [OSVR/OSVR-Core#128]
An in-process server would be a useful addition for the simple and very common use case - a single application accessing local devices that are not shared with multiple applications/users at the same time. In that case a separate server process that the application talks to through a socket is an overkill. 

Requiring an external server process (whether co-located with the client or remote) is frequently problematic for the following reasons:

* The external server process needs to be somehow managed by the application that uses it - started, stopped, restarted, potentially reconfigured, reset, calibration triggered, etc. As it is a separate process, this often needs to be done by "out of band" means, making integration into the application's UI difficult and complicated. This is an unfortunate trait that is shared with the VRPN framework. 

* Separate networked process can be a major support issue due to problems with antivirus/firewall software interfering with the network connections, TCP/UDP port conflicts, etc.

* Some platforms, such as Android, do not allow to run separate processes easily. This needs to be managed by the means of an Android service, both the client app and the service must have correct permissions otherwise they will not be able to communicate with each other, etc.

* The extra service running on the mobile device means additional battery drain.

##### Proposed enhancement
I suggest adding an option to have a lightweight in-process server class, bypassing the client-server networking. This *should not* be seen as a replacement for the standalone server but as an alternative for the smaller (but very common) use cases. Ideally the application should be able to choose when creating the OSVR context whether it wants to connect to its local (in-process) server or to a standalone (possibly remote) one - both should be possible.

**Comment by JeroMiya** [comment]
I was going to suggest this as well. I think it's an important use case. Most games/apps on the consumer side will only be connecting to localhost, so allowing an integrated OSVR server in the same process as the client kit would make OSVR much easier to use for the most common/simple scenario.

As mentioned, it also enables OSVR to be used on platforms where spawning a separate process or connecting to localhost is prohibited by the application sandbox. For example, windows store apps are prohibited from connecting to localhost, and could only connect to an OSVR server running on another machine, which might mean that it couldn't work with a locally plugged in HMD.

That being said, the integrated OSVR server would still need to be able to connect to a remote OSVR server and mirror an interface from that server, or merge the local interface tree with the remote one. This would enable the use of input devices that can't connect to the mobile device or console (for example, using a leap motion with an android app). In that scenario, the user could connect the input device to another machine on their LAN and mirror it into the mobile app's OSVR server.

Could this be done by an abstraction around the transport layer of OSVR-Server (and ClientKit on the receiving end), such that you could pass it an in-memory stream rather than a TCP/IP or UDP socket?

**Comment by russell-taylor** [comment]
There is a VRPN loopback connection that doesn't use any networking at all.  It is not yet multi-threaded, and does not support logging.  It could be used as an example.

**Comment by rpavlik** [comment]
Yeah, this has been in consideration/designed for since the beginning. The catch is how to design the client API.

**Comment by janoc** [comment]
Well, the API should ideally remain the same, perhaps passing a flag that one wants an in-process server when creating the context. I would even see this being a default, as it will likely be the most common use case.

**Comment by JeroMiya** [comment]
Perhaps defaulting to creating an internal in-process server could work as a fallback. However, I think developers might need the opportunity to configure the in-process server before connecting a client context to it. Maybe an osvrInitClientWithServer function, passing in a server instance?

**Comment by rpavlik** [comment]
Yeah, we will need to careful design this, since not finding a server will be used as a proxy for "no vr hardware" by some apps.
